Create rule S6484: Concatenated "std::format" outputs should be replaced by a single invocation (CPP-3950) (#1440)
SonarSource/rspec: rules/S6484/cfamily/rule.adoc

`std::format` accepts a format string composed of ordinary text and replacement fields (surrounded with `{}`) that are replaced with a textual representation of the next `std::format` arguments.
This allows generating a complex string with a single invocation of `std::format`.

Since calls to `std::format` produce string objects, it is possible to concatenate them with other string objects or string literals.
However, compared to a single `std::format` invocation with an adjusted format string, this concatenation is inefficient and less readable.

This rule raises an issue when a concatenation an `std::format` invocation can be replaced with a simple `std::format` invocation.

== Noncompliant Code Example

[source,cpp]
----
void formatExamples(std::string str, char const* cstr, int i) {
  std::string s1 = "You have been greeted " + std::format("{}", i) + " times."; // Noncompliant
  std::string s2 = "Hello " + std::format("{:*^20}", str) + "! " + std::format("{:->15}", cstr) + '.'; // Noncompliant

>   std::string s3 =  "Welcome";   // Noncompliant, this and 4 next line can be merged

Reviewer: ```suggestion
  std::string s3 =  "Welcome";   // Noncompliant, this and the following four lines can be merged
```

`std::format` accepts a format string that in addition to replacement fields (surrounded with `{}`) which are replaced with a 
textual representation of the arguments, can include hardcoded text surrounding them. This allows to produce a complex output,
 with a single invocation of `std::format`.

As calls to `std::format` produce string objects, it is possible to concatenate them with other string objects or string literals.
However, compared to single `std::format` invocation with adjusted format string, such construct is less readable and efficient.

This rule raises an issue when a concatenation of the outpt of `std::format` can be replaced with single `std::format` invocation.

== Noncompliant Code Example

[source,cpp]
----
void formatExamples(std::string str, char const* cstr, int i) {

>   std::string s1 = "The int argument has '" + std::format("{}", i) + "' value."; // Noncompliant

Reviewer 2: ```suggestion
  std::string s1 = "The int argument has a value of '" + std::format("{}", i) + "'."; // Noncompliant
```

Reviewer 2: There is still one string on each side so it would still be valid. Another solution would be to have something like the other twos:
`"You have been greeted " + std::format("{}", i) + " times";`

  std::string s2 = "Hello " + std::format("{:*^20}", str) + "! " + std::format("{:->15}", cstr) + '.'; // Noncompliant
  std::string s3 =  "Welcome";   // Noncompliant, this and 4 next line can be merged
  s3 +=  std::format("{:*^20}", str);
  s3 += "! ";
  s3 += std::format("{:->15}", cstr);
  s3 += '.';
}

> std::string fullName(std::string name, std::string secondName, std::string surname, std::size_t number) {

Reviewer 2: Do we usually have such "Compliant" code inside the Noncompliant example, that is not repeated in the "Compliant Solution"?

Reviewer 3: That was always my approach. I can make it exception. What that be better?

Reviewer 2: If we have the same pattern in other places, no need to change it

  // Compliant, as the formatted output depends on runtime properties
  std::string result = std::format("({}) {}", number, name);
  if (!secondName.empty()) {
     result += " ";
     result += secondName.front();
  }
  result += surname;
  return result;
}
----

== Compliant Solution

[source,cpp]
----
void formatExamples(std::string str, char const* cstr, int i) {

>   std::string s1 = std::format("The int argument has '{}' value.", i); // Compliant

Reviewer 2: ```suggestion
  std::string s1 = std::format("The int argument has a value of '{}'.", i); // Compliant
```

Reviewer 3: As before.
